File: Solution/WebSolution/Java/AssessmentAPI/src/main/java/com/transflower/tflassessment/services/QuestionBankService.java
package com.transflower.tflassessment.services;

import java.util.List;
import java.util.concurrent.CompletableFuture;

import com.transflower.tflassessment.entities.NewQuestion;
import com.transflower.tflassessment.entities.Question;
import com.transflower.tflassessment.entities.QuestionDetails;
import com.transflower.tflassessment.entities.QuestionTitle;
import com.transflower.tflassessment.entities.SubjectQuestion;

public interface QuestionBankService  {

    public CompletableFuture<List<QuestionTitle>> getAllQuestions();

    public CompletableFuture<List<SubjectQuestion>> getQuestionsBySubject(int id);

    public CompletableFuture<List<QuestionDetails>> getQuestionsBySubjectAndConcept(int subjectId, int conceptId);

    public CompletableFuture<List<QuestionDetails>> getQuestionsWithSubjectAndConcept();

    public CompletableFuture<List<Question>> getQuestions(int testId);

    public CompletableFuture<Boolean> updateAnswer(int id, char answerKey);

    public CompletableFuture<Question> getQuestion(int questionId);

    public CompletableFuture<Boolean> updateQuestionOptions(int id, Question options);

    public CompletableFuture<Boolean> updateSubjectConcept(int questionId, Question question);

    public CompletableFuture<Boolean> insertQuestion(NewQuestion question);

    public CompletableFuture<String> getConcept(String subject, int questionId);

}

File: Solution/WebSolution/Dotnet/AssessmentPortal/Monolithic/Monolithic_Controller/AssessmentServices/Implementations/QuestionBankService.cs
using Transflower.TFLAssessment.Entities;
using Transflower.TFLAssessment.Services.Interfaces;
using  Transflower.TFLAssessment.Repositories.Interfaces;
namespace Transflower.TFLAssessment.Services;

public class QuestionBankService:IQuestionBankService
{
    private readonly IQuestionBankRepository _repository;
    public QuestionBankService(IQuestionBankRepository repository)
    {
        _repository = repository;
    }

    public async Task<List<QuestionTitle>> GetAllQuestions(){
      return await _repository.GetAllQuestions();
    }

    public async Task<List<SubjectQuestion>> GetQuestionsBySubject(int id)
    {
        return await _repository.GetQuestionsBySubject(id);
        
    }

    public async Task<List<QuestionDetails>> GetQuestionsBySubjectAndConcept(int subjectId,int conceptId)
    {
        return await _repository.GetQuestionsBySubjectAndConcept(subjectId, conceptId);
       
    }

      public async Task<List<QuestionDetails>> GetQuestionsWithSubjectAndConcept()
    {
        return await _repository.GetQuestionsWithSubjectAndConcept();
       
    }

    public async Task<bool> UpdateAnswer(int id, char answerKey){
       return await _repository.UpdateAnswer(id,answerKey);
    }

    public async Task<Question> GetQuestion(int questionId)
    {   
        return await _repository.GetQuestion(questionId);
    }

    public async Task<List<Question>> GetQuestions(int testId)
    { 
       return await _repository.GetQuestions(testId);
    }

    public async Task<bool> UpdateQuestionOptions(int id,Question options)
    { 
       return await _repository.UpdateQuestionOptions(id,options);
      
    }


    //update only evaluationcriteriaid
    public async Task<bool> UpdateSubjectConcept(int questionId,Question question)
    {
        return await _repository.UpdateSubjectConcept(questionId,question);
        
       
    }


    public async Task<bool> InsertQuestion(NewQuestion question)
    {
        if (question == null)
        {
            throw new ArgumentNullException(nameof(question), "Question cannot be null");
        }
       return await _repository.InsertQuestion(question);
    }

    
    public async Task<string> GetConcept(string subject, int questionId)
    {
        return await _repository.GetConcept(subject,questionId);
    }


    public async Task<List<SubjectQuestionCount>> GetSubjectQuestionCount()
    {
        return await _repository.GetSubjectQuestionCount();
    }
}

